cross: regression with armv7 pkgsStatic (error: C compiler cannot create executables)

**Describe the bug**
With #101606 many `pkgsCross.armv7l-hf-multiplatform.pkgsStatic.` attributes that were working stopped building.

**To Reproduce**
```
 $ env -i nix-build -A pkgsCross.armv7l-hf-multiplatform.pkgsStatic.hello
[...]
configure: error: in `/build/hello-2.10':                                                                                                    
configure: error: C compiler cannot create executables                                                                                       
See `config.log' for more details                                                                                                            
builder for '/nix/store/ik89nsfcnjwrb5vz7chwc2accp604grh-hello-2.10-armv7l-unknown-linux-musleabihf.drv' failed with exit code 77            
```

**Additional context**
Reverting 4e9dc46dea0ef8cf15c567fa863796bb23099d0b makes many thing work again.

I'm not sure what should be done here. On one hand I *assume* PIE is desired. But it definitely provably breaks things with cross-compilation.

It seems to be limited to armv7; the equivalent aarch64 builds work fine.

**Notify maintainers**
@Ericson2314 @matthewbauer as cross-compilation aficionados
@utsl42 author of the change

Relevant `config.log` excerpt:

```
configure:4014: checking whether the C compiler works
configure:4036: armv7l-unknown-linux-musleabihf-gcc    conftest.c  >&5
/nix/store/lxzkkscp52ighr658k33wl71749rf8sk-armv7l-unknown-linux-musleabihf-binutils-2.35.1/bin/armv7l-unknown-linux-musleabihf-ld: /nix/store/8939rxc7d16hh25cxbcgym0pk9202s2a-armv7l-unknown-linux-musleabihf-stage-final-gcc-debug-10.2.0/lib/gcc/armv7l-unknown-linux-musleabihf/10.2.0/crtbeginT.o: relocation R_ARM_MOVW_ABS_NC against `a local symbol' can not be used when making a shared object; recompile with -fPIC
/nix/store/8939rxc7d16hh25cxbcgym0pk9202s2a-armv7l-unknown-linux-musleabihf-stage-final-gcc-debug-10.2.0/lib/gcc/armv7l-unknown-linux-musleabihf/10.2.0/crtbeginT.o:(.init_array+0x0): dangerous relocation: unsupported relocation
/nix/store/8939rxc7d16hh25cxbcgym0pk9202s2a-armv7l-unknown-linux-musleabihf-stage-final-gcc-debug-10.2.0/lib/gcc/armv7l-unknown-linux-musleabihf/10.2.0/crtbeginT.o:(.fini_array+0x0): dangerous relocation: unsupported relocation
collect2: error: ld returned 1 exit status
configure:4040: $? = 1
configure:4078: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "GNU Hello"
| #define PACKAGE_TARNAME "hello"
| #define PACKAGE_VERSION "2.10"
| #define PACKAGE_STRING "GNU Hello 2.10"
| #define PACKAGE_BUGREPORT "bug-hello@gnu.org"
| #define PACKAGE_URL "http://www.gnu.org/software/hello/"
| #define PACKAGE "hello"
| #define VERSION "2.10"
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:4083: error: in `/build/hello-2.10':
configure:4085: error: C compiler cannot create executables
```

--- a/pkgs/stdenv/generic/make-derivation.nix
+++ b/pkgs/stdenv/generic/make-derivation.nix
@@ -106,7 +106,12 @@ in rec {
                                       ++ depsTargetTarget ++ depsTargetTargetPropagated) == 0;
       dontAddHostSuffix = attrs ? outputHash && !noNonNativeDeps || (stdenv.noCC or false);
       supportedHardeningFlags = [ "fortify" "stackprotector" "pie" "pic" "strictoverflow" "format" "relro" "bindnow" ];
-      defaultHardeningFlags = if stdenv.hostPlatform.isMusl
+                              # Musl-based platforms will keep "pie", other platforms will not.
+      defaultHardeningFlags = if stdenv.hostPlatform.isMusl &&
+                                # Except when:
+                                #    - static aarch64, where compilation works, but produces segfaulting dynamically linked binaries.
+                                #    - static armv7l, where compilation fails.
+                                !((stdenv.hostPlatform.isAarch64 || stdenv.hostPlatform.isAarch32) && stdenv.hostPlatform.isStatic)
                               then supportedHardeningFlags
                               else lib.remove "pie" supportedHardeningFlags;
       enabledHardeningOptions =

